### PR TITLE
feat: Working state for `conversations` stream

### LINKED ISSUE #17: Implement State with the Conversations stream
@visch seems like the only way to limit what's fetched is via the offset and limit.

I did see this in the documentation `Retrieves all conversations sorted by the lastReceivedDatetime` - so maybe there's some filtering on the tap side that could happen in a separate iteration.

_Originally posted by @tayloramurphy in https://github.com/MeltanoLabs/tap-messagebird/issues/16#issuecomment-1399015938_

### PR DESCRIPTION
**Pull request by visch**
Closes #17

### CODE REVIEW

File: tap_messagebird/streams.py [hunk 1]
```diff
@@ -156,9 +187,12 @@ def get_url_params(
...
-            params["from"] = self.config["start_date"]
+            from_date = pendulum.parse(self.config["start_date"]).to_iso8601_string()
+            params["from"] = from_date
         return params
 
 
 class ConversationArchivedWarning(Exception):
     """Conversation is archived and we recieved an error."""
+
+    """Conversation is archived and we recieved an error."""
```
Comment (visch) [inline tap_messagebird/streams.py:198]: whoops, fixing

File: tap_messagebird/streams.py [hunk 2]
```diff
@@ -84,20 +127,8 @@ class ConversationMessagesStream(MessagebirdConversations):
     # Optionally, you may also use `schema_filepath` in place of `schema`:
     schema_filepath = SCHEMAS_DIR / "conversation_message.json"
     parent_stream_type = ConversationsStream
-
-    def get_url_params(
```
Comment (visch) [inline tap_messagebird/streams.py:88]: From looking at produciton data, most conversations don't have a ton of back and forth so also keeping a state tracker for number of messages is probably overboard. We could but I wanted to get rid of the complexity and potential for something messing up

File: tap_messagebird/streams.py [hunk 3]
```diff
@@ -70,11 +72,52 @@ def get_child_context(
...
+                pendulum.DateTime,
+                pendulum.parse(record[self.replication_key]),
+            )
+            # Conversations are returned in descending order, so we can stop
+            # There's no filtering paramater just this default ordering
+            if record_last_received_datetime < starting_replication_key_value:
+                self.logger.info(
+                    f"Hit a record with a {str(record_last_received_datetime)=}"
+                    f"< {str(starting_replication_key_value)=}. Stopping.",
+                )
```
Comment (edgarrmondragon) [inline tap_messagebird/streams.py:111]:
```suggestion
                self.logger.info(
                    "Breaking after hitting a record with replication key %s < %s",
                    record_last_received_datetime,
                    starting_replication_key_value,
                )
```

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]
Would this use case be covered by https://github.com/meltano/sdk/issues/226?

**Comment by visch** [comment]
> Would this use case be covered by [meltano/sdk#226](https://github.com/meltano/sdk/issues/226)?

Kind of, I think instead of "faking" it though we are actually reducing the number of requests dramatically as we stop iterating through pagination. So maybe that feature would solve this I guess it depends on how it's implemented!

File: tap_messagebird/streams.py (hunk 3 above)
Comment (edgarrmondragon) [reply tap_messagebird/streams.py]: It's not hate 😅, just convention and improves performance without a large dev cost when filtering out logs by level

**Review by edgarrmondragon — APPROVED** [review]

**Pull request merged**